platform/linux-generic/odp_crypto.c

```diff
@@ -165,24 +192,51 @@ null_crypto_routine(odp_crypto_op_param_t *param ODP_UNUSED,
 	return ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+/* Mimic new OpenSSL 1.1.y API */
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+static HMAC_CTX *HMAC_CTX_new(void)
+{
+	HMAC_CTX *ctx = malloc(sizeof(*ctx));
+
+	HMAC_CTX_init(ctx);
+	return ctx;
+}
+
+static void HMAC_CTX_free(HMAC_CTX *ctx)
+{
+	HMAC_CTX_cleanup(ctx);
+	free(ctx);
+}
+#endif
+
 static
-void packet_hmac_calculate(HMAC_CTX *ctx,
-			   odp_crypto_op_param_t *param,
-			   odp_crypto_generic_session_t *session,
-			   uint8_t *hash)
+void packet_hmac(odp_crypto_op_param_t *param,
+		 odp_crypto_generic_session_t *session,
+		 uint8_t *hash)
 {
+	HMAC_CTX *ctx = local.ctx[session->idx].hmac;
 	odp_packet_t pkt = param->out_pkt;
 	uint32_t offset = param->auth_range.offset;
 	uint32_t len   = param->auth_range.length;
 
 	ODP_ASSERT(offset + len <= odp_packet_len(pkt));
 
-	HMAC_Init_ex(ctx,
-		     session->auth.key,
-		     session->auth.key_length,
-		     session->auth.evp_md,
-		     NULL);
+	if (!IS_VALID(session, hmac, local.id)) {
+		SET_VALID(session, hmac, local.id);
+		HMAC_Init_ex(ctx,
+			     session->auth.key,
+			     session->auth.key_length,
+			     session->auth.evp_md,
+			     NULL);
+	} else {
+		HMAC_Init_ex(ctx,
+			     NULL,
+			     session->auth.key_length,
```

Comment:
Why is this needed here? The key length was already set. If this is because of some OpenSSL weirdness, an explanatory comment would be nice.

Reply:
OK, I will add a comment. This is necessary to restart (reset) HMAC calculation without resetting the key and auth length.

Reply:
Yes, the hmac reset of course (what was I thinking?). But for that this should be enough (I have used it in my test patch with OpenSSL 1.1.0 and 1.0.2): HMAC_Init_ex(ctx, NULL, 0, NULL, NULL) and be more clear in that key len is not being changed.

Looking at the man page I do not think HMAC_Init_Ex() is going to even look at key_len when key is NULL.